webrender/src/render_backend.rs

```diff
@@ -470,6 +481,38 @@ impl RenderBackend {
         let mut notifier = self.notifier.lock();
         notifier.as_mut().unwrap().as_mut().unwrap().new_scroll_frame_ready(composite_needed);
     }
+
+    fn handle_vr_compositor_command(&mut self, ctx_id: WebGLContextId, cmd: VRCompositorCommand) {
+        match cmd {
+            VRCompositorCommand::Create(compositor_id) => {
+                let mut creator = self.vr_compositor_creator.lock();
+                let compositor = creator.as_mut().unwrap().as_mut().unwrap().create_compositor(compositor_id);
+                if let Some(compositor) = compositor {
+                    self.vr_compositors.insert(compositor_id, compositor);
+                }
+            }
+            VRCompositorCommand::SyncPoses(compositor_id, near, far, sender) => {
+                if let Some(ref compositor) = self.vr_compositors.get(&compositor_id) {
+                    let pose = compositor.sync_poses(near, far);
+                    let _result = sender.send(Ok(pose));
+                } else {
+                    let _result = sender.send(Err(()));
+                }
+            }
+            VRCompositorCommand::SubmitFrame(compositor_id, left_bounds, right_bounds) => {
+                if let Some(ref compositor) = self.vr_compositors.get(&compositor_id) {
+                    let texture = self.resource_cache.get_webgl_texture(&ctx_id);
```

Comment:
Besides the resource_cache access to get the current texture_id (which might change after a webgl resize) another requirement is that specified VRCompositor calls must be run in the same thread as the WebGL commands. There could be many different VRCompositor implementations for different VR vendors, and some of them require to be created/initialized in the render thread too (because they use the shared OpenGL texture and the active OpenGL context).

These functions must run in response to some ipc-channel messages from JavaScript and we have to minimize the render path to achieve WebVR latency requirements. There is no way in webrender to run generic closures because of serde-serialization and sandboxed processes, so I implemented it using the traits. I'm about to publish a blogpost with more info about the overall WebVR implementation architecture, I'll send you the link. If you have some ideas to narrow down this interface while keeping the latency requirements that would be great.